// @remixproject/engine-vscode/lib/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionPlugin = void 0;
const tslib_1 = require("tslib");
// Check here : https://code.visualstudio.com/api/references/vscode-api#extensions
const engine_1 = require("@remixproject/engine");
const vscode_1 = require("vscode");
class ExtensionPlugin extends engine_1.PluginConnector {
    constructor(profile) {
        super(profile);
    }
    send(message) {
        if (this.extension) {
            this.connector.onMessage(message);
        }
    }
    connect(url) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                this.extension = vscode_1.extensions.getExtension(url);
                this.connector = yield this.extension.activate();
            }
            catch (err) {
                throw new Error(`ExtensionPlugin "${this.profile.name}" could not connect to the engine.`);
            }
        });
    }
    disconnect() {
        if (this.extension) { }
    }
}
exports.ExtensionPlugin = ExtensionPlugin;
