// @remixproject/engine-vscode/lib/window.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WindowPlugin = exports.windowProfile = void 0;
const engine_1 = require("@remixproject/engine");
const vscode_1 = require("vscode");
exports.windowProfile = {
    name: 'window',
    methods: ['prompt', 'select', 'selectFile', 'selectFolder', 'alert', 'error', 'warning'],
};
class WindowPlugin extends engine_1.Plugin {
    constructor(options = {}) {
        super(exports.windowProfile);
        // Leave 1min to let the user interact with the window
        super.setOptions(Object.assign({ queueTimeout: 60000 }, options));
    }
    prompt(options) {
        return vscode_1.window.showInputBox(options);
    }
    select(items, options) {
        return vscode_1.window.showQuickPick(items, options);
    }
    selectFile() {
        return vscode_1.window.showOpenDialog({ canSelectFiles: true }).then(([file]) => file.fsPath);
    }
    selectFolder() {
        return vscode_1.window.showOpenDialog({ canSelectFolders: true }).then(([folder]) => folder.fsPath);
    }
    alert(message, actions = []) {
        return vscode_1.window.showInformationMessage(message, ...actions);
    }
    error(message, actions = []) {
        return vscode_1.window.showErrorMessage(message, ...actions);
    }
    warning(message, actions = []) {
        return vscode_1.window.showWarningMessage(message, ...actions);
    }
}
exports.WindowPlugin = WindowPlugin;
