// @remix-project/remix-debug/src/index.js
'use strict';
const tslib_1 = require("tslib");
const init = tslib_1.__importStar(require("./init"));
const Ethdebugger_1 = require("./Ethdebugger");
const debugger_1 = require("./debugger/debugger");
const cmdline_1 = require("./cmdline");
const storageViewer_1 = require("./storage/storageViewer");
const storageResolver_1 = require("./storage/storageResolver");
const SolidityDecoder = tslib_1.__importStar(require("./solidity-decoder"));
const breakpointManager_1 = require("./code/breakpointManager");
const sourceMappingDecoder = tslib_1.__importStar(require("./source/sourceMappingDecoder"));
const traceHelper = tslib_1.__importStar(require("./trace/traceHelper"));
module.exports = {
    init,
    traceHelper,
    sourceMappingDecoder,
    EthDebugger: Ethdebugger_1.Ethdebugger,
    TransactionDebugger: debugger_1.Debugger,
    /**
     * constructor
     *
     * @param {Object} _debugger - type of EthDebugger
     * @return {Function} _locationToRowConverter - function implemented by editor which return a column/line position for a char source location
     */
    BreakpointManager: breakpointManager_1.BreakpointManager,
    SolidityDecoder,
    storage: {
        StorageViewer: storageViewer_1.StorageViewer,
        StorageResolver: storageResolver_1.StorageResolver
    },
    CmdLine: cmdline_1.CmdLine
};
